File: go/test/output_list_of_available_role_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/bluefin-tecs-user-backoffice-sdk/go"
	"github.com/voxgig-sdk/bluefin-tecs-user-backoffice-sdk/go/core"

	vs "github.com/voxgig-sdk/bluefin-tecs-user-backoffice-sdk/go/utility/struct"
)

func TestOutputListOfAvailableRoleEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.OutputListOfAvailableRole(nil)
		if ent == nil {
			t.Fatal("expected non-nil OutputListOfAvailableRoleEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := output_list_of_available_roleBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "output_list_of_available_role." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set BLUEFINTECSUSERBACKOFFICE_TEST_OUTPUT_LIST_OF_AVAILABLE_ROLE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		outputListOfAvailableRoleRef01Ent := client.OutputListOfAvailableRole(nil)
		outputListOfAvailableRoleRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "output_list_of_available_role"}, setup.data), "output_list_of_available_role_ref01"))

		outputListOfAvailableRoleRef01DataResult, err := outputListOfAvailableRoleRef01Ent.Create(outputListOfAvailableRoleRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		outputListOfAvailableRoleRef01Data = core.ToMapAny(outputListOfAvailableRoleRef01DataResult)
		if outputListOfAvailableRoleRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

	})
}

func output_list_of_available_roleBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "output_list_of_available_role", "OutputListOfAvailableRoleTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read output_list_of_available_role test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse output_list_of_available_role test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"output_list_of_available_role01", "output_list_of_available_role02", "output_list_of_available_role03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("BLUEFINTECSUSERBACKOFFICE_TEST_OUTPUT_LIST_OF_AVAILABLE_ROLE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"BLUEFINTECSUSERBACKOFFICE_TEST_OUTPUT_LIST_OF_AVAILABLE_ROLE_ENTID": idmap,
		"BLUEFINTECSUSERBACKOFFICE_TEST_LIVE":      "FALSE",
		"BLUEFINTECSUSERBACKOFFICE_TEST_EXPLAIN":   "FALSE",
		"BLUEFINTECSUSERBACKOFFICE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["BLUEFINTECSUSERBACKOFFICE_TEST_OUTPUT_LIST_OF_AVAILABLE_ROLE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["BLUEFINTECSUSERBACKOFFICE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["BLUEFINTECSUSERBACKOFFICE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewBluefinTecsUserBackofficeSDK(core.ToMapAny(mergedOpts))
	}

	live := env["BLUEFINTECSUSERBACKOFFICE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["BLUEFINTECSUSERBACKOFFICE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: cpp/test/output_list_of_available_role_entity_test.cpp
// Generated basic-flow test for the output_list_of_available_role entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct OutputListOfAvailableRoleSetup {
  std::shared_ptr<BluefinTecsUserBackofficeSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static OutputListOfAvailableRoleSetup output_list_of_available_role_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/output_list_of_available_role/OutputListOfAvailableRoleTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinTecsUserBackofficeSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("output_list_of_available_role01"), Value("output_list_of_available_role02"), Value("output_list_of_available_role03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINTECSUSERBACKOFFICE_TEST_OUTPUT_LIST_OF_AVAILABLE_ROLE_ENTID", idmap},
    {"BLUEFINTECSUSERBACKOFFICE_TEST_LIVE", Value("FALSE")},
    {"BLUEFINTECSUSERBACKOFFICE_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINTECSUSERBACKOFFICE_TEST_OUTPUT_LIST_OF_AVAILABLE_ROLE_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINTECSUSERBACKOFFICE_TEST_LIVE") == Value("TRUE");

  OutputListOfAvailableRoleSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void output_list_of_available_role_entity_instance() {
  auto testsdk = BluefinTecsUserBackofficeSDK::testSDK();
  auto ent = testsdk->output_list_of_available_role();
  ASSERT_EQ(ent->getName(), std::string("output_list_of_available_role"), "entity name");
}

static void output_list_of_available_role_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"output_list_of_available_role", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinTecsUserBackofficeSDK::testSDK(seed, sdkopts);
  auto se = strsdk->output_list_of_available_role();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinTecsUserBackofficeSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->output_list_of_available_role();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void output_list_of_available_role_entity_basic() {
  auto setup = output_list_of_available_role_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create"}) {
    auto sk = is_control_skipped("entityOp", std::string("output_list_of_available_role.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto output_list_of_available_role_ref01_ent = client->output_list_of_available_role();
  Value output_list_of_available_role_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "output_list_of_available_role"}), "output_list_of_available_role_ref01"));
  if (!output_list_of_available_role_ref01_data.is_map()) output_list_of_available_role_ref01_data = vmap();
  {
    Value output_list_of_available_role_ref01_data_result = output_list_of_available_role_ref01_ent->create(Struct::clone(output_list_of_available_role_ref01_data), Value::undef());
    output_list_of_available_role_ref01_data = Helpers::toMapAny(output_list_of_available_role_ref01_data_result);
    if (!output_list_of_available_role_ref01_data.is_map()) output_list_of_available_role_ref01_data = vmap();
    ASSERT_TRUE(output_list_of_available_role_ref01_data.is_map(), "expected create result to be a map");
  }

}

int main() {
  T_RUN(output_list_of_available_role_entity_instance);
  T_RUN(output_list_of_available_role_entity_stream);
  T_RUN(output_list_of_available_role_entity_basic);
  return sdktest::summary("output_list_of_available_role_entity_test");
}
